### ISSUE TITLE
Add some util functions into lexlib

### ISSUE THREAD

**Issue by zufuliu** [ScintillaOrg/lexilla#17]
1. AnyOf(ch, ...), source code at https://github.com/zufuliu/notepad2/blob/master/scintilla/lexlib/CharacterSet.h#L105
It has no runtime setup cost (compiled into bit test or lookup table).
2. Move implementation for GetRange() and GetRangeLowered() from StyleContext into LexAccessor, source code at https://github.com/zufuliu/notepad2/blob/master/scintilla/lexlib/LexAccessor.cxx#L30
Many lexers (HTML, Ruby, lexer using lineBuffer, etc.) can change to the new methods.
3. MatchIgnoreCase().

[LexAccessor.zip](https://github.com/ScintillaOrg/lexilla/files/6721525/LexAccessor.zip)

**Comment by zufuliu** [comment]
update MatchIgnoreCase:
```cpp
bool LexAccessor::MatchIgnoreCase(Sci_Position pos, const char *s) noexcept {
	for (; *s; s++, pos++) {
		if (*s != MakeLowerCase(SafeGetCharAt(pos, '\0'))) {
			return false;
		}
	}
	return true;
}
```

**Comment by zufuliu** [comment]
There are buffer overrun in GetRange() and GetRangeLowered() :
```cpp
assert(startPos_ <= endPos_ && len != 0 && s != nullptr);
endPos_ = sci::min(endPos_, startPos_ + len - 1);
```

**Comment by nyamatongwe** [comment]
> Move implementation for GetRange() and GetRangeLowered()

These are not implemented in the distributed version of StyleContext. Perhaps you mean GetCurrent?

**Comment by zufuliu** [comment]
Yes, changes for StyleContext.h (moved the implementation from StyleContext.cxx into LexAccessor):
```cpp
void GetCurrent(char *s, Sci_PositionU len) {
	styler.GetRange(styler.GetStartSegment(), currentPos, s, len);
}
void GetCurrentLowered(char *s, Sci_PositionU len) {
	styler.GetRangeLowered(styler.GetStartSegment(), currentPos, s, len);
}

```

**Comment by nyamatongwe** [comment]
AnyOf added - that's quite cute and the generated code changes a lot depending on the actual values.

Shouldn't GetRangeLowered call GetRange for most of its implementation to avoid duplicated code?

**Comment by zufuliu** [comment]
Changed to call GetRange().
[LexAccessor-2.zip](https://github.com/ScintillaOrg/lexilla/files/6755117/LexAccessor-2.zip)

**Comment by nyamatongwe** [comment]
Committed as 2b36c2ed3491838acb580d4ba52bd20eac2e4aea with some minor changes: fix inconsistent noexcept; hid StyleContext methods back in implementation file; added to build files.

**Comment by nyamatongwe** [comment]
There is a size mismatch in the string returning `LexAccessor::GetRange` and `LexAccessor::GetRangeLowered` methods leading to removing the last character in the range.

~~~c++
// Changing:
char s[100];
styler.GetRange(styler.GetStartSegment(), sc.currentPos, s, sizeof(s));
// to
const std::string st = styler.GetRange(styler.GetStartSegment(), sc.currentPos);
const char *s = st.c_str();
~~~

The problem is that `LexAccessor::GetRange` isn't allowing for the ending NUL. It correctly allocates a `std::string` of size `len` with a NUL after the `len` but the called `LexAccessor::GetRange` further restricts the length to `startPos_ + len - 1 - startPos_` or `len - 1`. It is safe to overwrite the terminating NUL in a `std::string` so this change could be made:
~~~patch
diff --git a/lexlib/LexAccessor.cxx b/lexlib/LexAccessor.cxx
index 0229e475..f0d6bdf8 100644
--- a/lexlib/LexAccessor.cxx
+++ b/lexlib/LexAccessor.cxx
@@ -58,7 +58,7 @@ std::string LexAccessor::GetRange(Sci_PositionU startPos_, Sci_PositionU endPos_
 	assert(startPos_ < endPos_);
 	const Sci_PositionU len = endPos_ - startPos_;
 	std::string s(len, '\0');
-	GetRange(startPos_, endPos_, s.data(), len);
+	GetRange(startPos_, endPos_, s.data(), len + 1);
 	return s;
 }
 
@@ -66,7 +66,7 @@ std::string LexAccessor::GetRangeLowered(Sci_PositionU startPos_, Sci_PositionU
 	assert(startPos_ < endPos_);
 	const Sci_PositionU len = endPos_ - startPos_;
 	std::string s(len, '\0');
-	GetRangeLowered(startPos_, endPos_, s.data(), len);
+	GetRangeLowered(startPos_, endPos_, s.data(), len + 1);
 	return s;
 }
~~~

**Comment by zufuliu** [comment]
Oh, yes.

**Comment by nyamatongwe** [comment]
Included in 5.2.0.
